### ISSUE TITLE
Treat arrays as list

### ISSUE THREAD

**Issue by glebm** [vishesh/racketscript#132]
If arrays were treated as lists, we could:

1. Get rid of the distinction between lambda and v-lambda.
2. Use spread arguments for varargs everywhere (#102).
3. Interop with JavaScript more easily.

The implementation for most operations can be made very simple if we implement `Symbol.iterator` for `Pair`.

**Comment by vishesh** [comment]
Lists are pair of two values and persistent. You can have some other datatype pointing at middle of the list. When you do `cdr` it will involve popping the first item (which we can't do as lists are immutable), so you would either want to make another copy or low-level representation of pointer to that array. This all sounds rather complicated, may actually make FFI more confusing.

Regarding, implementing Pair as iterator, I think we talked and I think its a good idea. We should do that for Vectors/Set/Hash as well. However, since pairs are not always lists, it may not make sense to always iterate them.

**Comment by glebm** [comment]
> When you do cdr it will involve popping the first item (which we can't do as lists are immutable), so you would either want to make another copy or low-level representation of pointer to that array.

Arrays of reasonable size (under 100k elements) are copy-on-write, so `cdr` could just be `slice(1)` and work in constant time (because lists are immutable so we never write). I'll write some benchmarks.

Aside: I'm not so sure myself that this will turn out to be a bright idea but still let's discuss.

**Comment by vishesh** [comment]
Cool. We definitely want to explore alternative encodings. Just warning that this probably wont end up in codebase in matter of days or even weeks. We'll need extensive tests, benchmarks (both micro/macro benchmarks) across browsers (at-least Firefox and Safari I think). Secondly, we'll have to think about what it means to move these across Racket <-> JS boundary.

**Comment by stchang** [comment]
Bumping this because it would be so nice to have racket `apply` just be js `apply`.

I suspect that actually implementing this is much more complex than we realize though, and that there many (corner) cases we are not thinking of that would make this not work, but I would be really interested if someone investigated this, a student project maybe?

**Comment by vishesh** [comment]
I think it would be interesting. I think it we'd easily break stuff that depends on lower level things like `eqv?`:

```racket
> (define a '(1 2 3))
#t
> (eqv? (cdr (cons 4 a)) (cdr (cons 4 a))) 
#t ;; with this change will likely be #f as it would copy the rest after doing cons
```

**Comment by stchang** [comment]
I agree, I dont think it would ultimately work. It would be so nice though.
